File: Harish Kumar/Array and Strings/Magnetic Force Between Two Balls/Magnetic Force Between Two Balls.java
class Solution {
    public boolean isAnswer(int[] arr, int mid, int m)
    {
        int count = 1;
        int prev = arr[0];
        
        for(int i = 1; i < arr.length; i++)
        {
            if(arr[i] - prev >= mid)
            {
                prev = arr[i];
                count++;
            }
        }
        return count >= m;
    }
    public int maxDistance(int[] position, int m) {
        
        Arrays.sort(position);
        
        int start = 0;
        int end = position[position.length-1];
        
        int ans = 0;
        while(start <= end)
        {
            int mid = start + (end - start) / 2;
            
            if(isAnswer(position, mid, m))
            {
                ans = mid;
                start = mid +1;
            }
            else
            {
                end = mid -1;
            }
        }
        return ans;
    }
}

File: Ankit Arora/Arrays and Strings/MagneticForceBetweentwoballs.cpp
class Solution {
public:
    bool isAnswer(vector<int>&position , int mid , int m)
    {
        int initial = 1;
        int prev = position[0];
        for(int i=1 ; i<position.size() ; i++)
        {
            if(position[i]-prev >= mid)
            {
                initial++;
                prev = position[i];
                if(initial==m) return true;
            }
        }
        if(initial<m) return false;
        return true;
    }


    int maxDistance(vector<int>& position, int m) {
        sort(position.begin(),position.end());
        int start = 0 , n = position.size(), end = position[n-1];
        int ans;
        while(start<=end)
        {
            int mid = start + (end - start)/2;
            if(isAnswer(position,mid,m))
            {
                ans = mid;
                start = mid + 1;
            }
            else
            {
                end = mid - 1;
            }
        }
        return ans;

    }
};

